// @uiw/codemirror-themes/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.createTheme = void 0;

var _view = require("@codemirror/view");

var _language = require("@codemirror/language");

var createTheme = function createTheme(_ref) {
  var theme = _ref.theme,
      settings = _ref.settings,
      styles = _ref.styles;
  var themeOptions = {
    '&': {
      backgroundColor: settings.background,
      color: settings.foreground
    },
    '.cm-gutters': {}
  };

  if (settings.gutterBackground) {
    themeOptions['.cm-gutters'].backgroundColor = settings.gutterBackground;
  }

  if (settings.gutterForeground) {
    themeOptions['.cm-gutters'].color = settings.gutterForeground;
  }

  if (settings.gutterBorder) {
    themeOptions['.cm-gutters'].borderRightColor = settings.gutterBorder;
  }

  if (settings.caret) {
    themeOptions['.cm-content'] = {
      caretColor: settings.caret
    };
    themeOptions['.cm-cursor, .cm-dropCursor'] = {
      borderLeftColor: settings.caret
    };
  }

  if (settings.lineHighlight) {
    themeOptions['.cm-activeLine'] = {
      backgroundColor: settings.lineHighlight
    };
    themeOptions['.cm-activeLineGutter'] = {
      backgroundColor: settings.lineHighlight
    };
  }

  if (settings.selection) {
    themeOptions['&.cm-focused .cm-selectionBackground .cm-selectionBackground, & .cm-selectionLayer .cm-selectionBackground, ::selection'] = {
      backgroundColor: settings.selection
    };
  }

  if (settings.selectionMatch) {
    themeOptions['& .cm-selectionMatch'] = {
      backgroundColor: settings.selectionMatch
    };
  }

  var themeExtension = _view.EditorView.theme(themeOptions, {
    dark: theme === 'dark'
  });

  var highlightStyle = _language.HighlightStyle.define(styles);

  var extension = [themeExtension, (0, _language.syntaxHighlighting)(highlightStyle)];
  return extension;
};

exports.createTheme = createTheme;
var _default = createTheme;
exports["default"] = _default;
