// @uiw/react-markdown-editor/cjs/commands/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getModeCommands = exports.getCommands = exports.defaultCommands = void 0;

var _bold = require("./bold");

var _italic = require("./italic");

var _header = require("./header");

var _strike = require("./strike");

var _underline = require("./underline");

var _olist = require("./olist");

var _ulist = require("./ulist");

var _quote = require("./quote");

var _link = require("./link");

var _todo = require("./todo");

var _image = require("./image");

var _redo = require("./redo");

var _undo = require("./undo");

var _fullscreen = require("./fullscreen");

var _preview = require("./preview");

var defaultCommands = {
  undo: _undo.undo,
  redo: _redo.redo,
  bold: _bold.bold,
  italic: _italic.italic,
  header: _header.header,
  strike: _strike.strike,
  underline: _underline.underline,
  quote: _quote.quote,
  olist: _olist.olist,
  ulist: _ulist.ulist,
  todo: _todo.todo,
  link: _link.link,
  image: _image.image,
  fullscreen: _fullscreen.fullscreen,
  preview: _preview.preview
};
exports.defaultCommands = defaultCommands;

var getCommands = function getCommands() {
  return Object.keys(defaultCommands).filter(function (key) {
    return !/^(fullscreen|preview)/.test(key);
  }).map(function (key) {
    return defaultCommands[key];
  });
};

exports.getCommands = getCommands;

var getModeCommands = function getModeCommands() {
  return [_preview.preview, _fullscreen.fullscreen];
};

exports.getModeCommands = getModeCommands;
